Behavior difference: tsgo wrongly allows `include` to contain files that is not under `rootDir` from tsconfig

## Steps to reproduce

1. Start a plain new project
    - "@typescript/native-preview": "7.0.0-dev.20260429.1",
    - "typescript": "6.0.3"
2. Run `npx tsgo --init` to get the default tsconfig. Turn on these flags as well.
    - "stableTypeOrdering": true,
    - "rootDir": "./src2",
    - "include": ["src1/*"]
3. Add an empty `src1/main.ts`. Run `npx tsc --noEmit` & `npx tsgo --noEmit` to see the difference.

## Behavior with `typescript@6.0`

```
error TS6059: File '<project>/src1/main.ts' is not under 'rootDir' '<project>/src2'. 'rootDir' is expected to contain all source files.
  The file is in the program because:
    Matched by include pattern 'src1/*' in '<project>/tsconfig.json'
  File is ECMAScript module because '<project>/package.json' has field "type" with value "module"

  tsconfig.json:40:17
    40     "include": ["src1/*"]
                       ~~~~~~~~
    File is matched by include pattern specified here.


Found 1 error.
```

## Behavior with `tsgo`

No error at all.

--- a/internal/compiler/emitter.go
+++ b/internal/compiler/emitter.go
@@ -480,7 +480,7 @@ func sourceFileMayBeEmitted(sourceFile *ast.SourceFile, host SourceFileMayBeEmit
 
 	// Otherwise, if rootDir is specified or a config file exists, we know the common source directory and can check if the file would be emitted in the same location
 	if options.RootDir != "" || options.ConfigFilePath != "" {
-		commonDir := tspath.GetNormalizedAbsolutePath(outputpaths.GetCommonSourceDirectory(options, nil, host.GetCurrentDirectory(), host.UseCaseSensitiveFileNames(), nil), host.GetCurrentDirectory())
+		commonDir := tspath.GetNormalizedAbsolutePath(outputpaths.GetCommonSourceDirectory(options, func() []string { return nil }, host.GetCurrentDirectory(), host.UseCaseSensitiveFileNames(), nil), host.GetCurrentDirectory())
 		outputPath := outputpaths.GetSourceFilePathInNewDirWorker(sourceFile.FileName(), options.OutDir, host.GetCurrentDirectory(), commonDir, host.UseCaseSensitiveFileNames())
 		if tspath.ComparePaths(sourceFile.FileName(), outputPath, tspath.ComparePathsOptions{
 			UseCaseSensitiveFileNames: host.UseCaseSensitiveFileNames(),

--- a/internal/compiler/emitter.go
+++ b/internal/compiler/emitter.go
@@ -480,7 +480,7 @@ func sourceFileMayBeEmitted(sourceFile *ast.SourceFile, host SourceFileMayBeEmit
 
 	// Otherwise, if rootDir is specified or a config file exists, we know the common source directory and can check if the file would be emitted in the same location
 	if options.RootDir != "" || options.ConfigFilePath != "" {
-		commonDir := tspath.GetNormalizedAbsolutePath(outputpaths.GetCommonSourceDirectory(options, func() []string { return nil }, host.GetCurrentDirectory(), host.UseCaseSensitiveFileNames()), host.GetCurrentDirectory())
+		commonDir := tspath.GetNormalizedAbsolutePath(outputpaths.GetCommonSourceDirectory(options, nil, host.GetCurrentDirectory(), host.UseCaseSensitiveFileNames(), nil), host.GetCurrentDirectory())
 		outputPath := outputpaths.GetSourceFilePathInNewDirWorker(sourceFile.FileName(), options.OutDir, host.GetCurrentDirectory(), commonDir, host.UseCaseSensitiveFileNames())
 		if tspath.ComparePaths(sourceFile.FileName(), outputPath, tspath.ComparePathsOptions{
 			UseCaseSensitiveFileNames: host.UseCaseSensitiveFileNames(),

--- a/internal/compiler/program.go
+++ b/internal/compiler/program.go
@@ -1534,22 +1534,39 @@ func (p *Program) GetDefaultLibFile(path tspath.Path) *LibFile {
 
 func (p *Program) CommonSourceDirectory() string {
 	p.commonSourceDirectoryOnce.Do(func() {
+		files := core.MapFiltered(p.files, func(file *ast.SourceFile) (string, bool) {
+			return file.FileName(), sourceFileMayBeEmitted(file, p, false /*forceDtsEmit*/) && !file.IsDeclarationFile
+		})
 		p.commonSourceDirectory = outputpaths.GetCommonSourceDirectory(
 			p.Options(),
-			func() []string {
-				var files []string
-				for _, file := range p.files {
-					if sourceFileMayBeEmitted(file, p, false /*forceDtsEmit*/) {
-						files = append(files, file.FileName())
-					}
-				}
-				return files
-			},
+			files,
 			p.GetCurrentDirectory(),
 			p.UseCaseSensitiveFileNames(),
+			p.checkSourceFilesBelongToPath,
 		)
 	})
 	return p.commonSourceDirectory
+}
+
+func (p *Program) checkSourceFilesBelongToPath(sourceFiles []string, rootDirectory string) bool {
+	allFilesBelongToPath := true
+	absoluteRootDirectoryPath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(rootDirectory, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
+	for _, file := range sourceFiles {
+		absoluteSourceFilePath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(file, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
+		if !strings.HasPrefix(absoluteSourceFilePath, absoluteRootDirectoryPath) {
+			p.includeProcessor.addProcessingDiagnostic(&processingDiagnostic{
+				kind: processingDiagnosticKindExplainingFileInclude,
+				data: &includeExplainingDiagnostic{
+					file:    tspath.Path(absoluteSourceFilePath),
+					message: diagnostics.File_0_is_not_under_rootDir_1_rootDir_is_expected_to_contain_all_source_files,
+					args:    []any{file, rootDirectory},
+				},
+			})
+			allFilesBelongToPath = false
+		}
+	}
+
+	return allFilesBelongToPath
 }
 
 type WriteFileData struct {

--- a/internal/compiler/program.go
+++ b/internal/compiler/program.go
@@ -1534,9 +1534,11 @@ func (p *Program) GetDefaultLibFile(path tspath.Path) *LibFile {
 
 func (p *Program) CommonSourceDirectory() string {
 	p.commonSourceDirectoryOnce.Do(func() {
-		files := core.MapFiltered(p.files, func(file *ast.SourceFile) (string, bool) {
-			return file.FileName(), sourceFileMayBeEmitted(file, p, false /*forceDtsEmit*/) && !file.IsDeclarationFile
-		})
+		files := func() []string {
+			return core.MapFiltered(p.files, func(file *ast.SourceFile) (string, bool) {
+				return file.FileName(), sourceFileMayBeEmitted(file, p, false /*forceDtsEmit*/) && !file.IsDeclarationFile
+			})
+		}
 		p.commonSourceDirectory = outputpaths.GetCommonSourceDirectory(
 			p.Options(),
 			files,

--- a/internal/compiler/program.go
+++ b/internal/compiler/program.go
@@ -1550,10 +1550,9 @@ func (p *Program) CommonSourceDirectory() string {
 
 func (p *Program) checkSourceFilesBelongToPath(sourceFiles []string, rootDirectory string) bool {
 	allFilesBelongToPath := true
-	absoluteRootDirectoryPath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(rootDirectory, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
 	for _, file := range sourceFiles {
 		absoluteSourceFilePath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(file, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
-		if !strings.HasPrefix(absoluteSourceFilePath, absoluteRootDirectoryPath) {
+		if !tspath.ContainsPath(rootDirectory, file, p.comparePathsOptions) {
 			p.includeProcessor.addProcessingDiagnostic(&processingDiagnostic{
 				kind: processingDiagnosticKindExplainingFileInclude,
 				data: &includeExplainingDiagnostic{

--- a/internal/outputpaths/commonsourcedirectory.go
+++ b/internal/outputpaths/commonsourcedirectory.go
@@ -56,16 +56,22 @@ func GetComputedCommonSourceDirectory(emittedFiles []string, currentDirectory st
 	return commonSourceDirectory
 }
 
-func GetCommonSourceDirectory(options *core.CompilerOptions, files func() []string, currentDirectory string, useCaseSensitiveFileNames bool) string {
+func GetCommonSourceDirectory(options *core.CompilerOptions, files []string, currentDirectory string, useCaseSensitiveFileNames bool, checkSourceFilesBelongToPath func([]string, string) bool) string {
 	var commonSourceDirectory string
 	if options.RootDir != "" {
 		// If a rootDir is specified use it as the commonSourceDirectory
 		commonSourceDirectory = options.RootDir
+		if checkSourceFilesBelongToPath != nil {
+			checkSourceFilesBelongToPath(files, options.RootDir)
+		}
 	} else if options.ConfigFilePath != "" {
 		// If the rootDir is not specified, then the common source directory is the directory of the config file.
 		commonSourceDirectory = tspath.GetDirectoryPath(options.ConfigFilePath)
+		if checkSourceFilesBelongToPath != nil {
+			checkSourceFilesBelongToPath(files, commonSourceDirectory)
+		}
 	} else {
-		commonSourceDirectory = computeCommonSourceDirectoryOfFilenames(files(), currentDirectory, useCaseSensitiveFileNames)
+		commonSourceDirectory = computeCommonSourceDirectoryOfFilenames(files, currentDirectory, useCaseSensitiveFileNames)
 	}
 
 	if len(commonSourceDirectory) > 0 {

--- a/internal/outputpaths/commonsourcedirectory.go
+++ b/internal/outputpaths/commonsourcedirectory.go
@@ -56,22 +56,22 @@ func GetComputedCommonSourceDirectory(emittedFiles []string, currentDirectory st
 	return commonSourceDirectory
 }
 
-func GetCommonSourceDirectory(options *core.CompilerOptions, files []string, currentDirectory string, useCaseSensitiveFileNames bool, checkSourceFilesBelongToPath func([]string, string) bool) string {
+func GetCommonSourceDirectory(options *core.CompilerOptions, files func() []string, currentDirectory string, useCaseSensitiveFileNames bool, checkSourceFilesBelongToPath func([]string, string) bool) string {
 	var commonSourceDirectory string
 	if options.RootDir != "" {
 		// If a rootDir is specified use it as the commonSourceDirectory
 		commonSourceDirectory = options.RootDir
 		if checkSourceFilesBelongToPath != nil {
-			checkSourceFilesBelongToPath(files, options.RootDir)
+			checkSourceFilesBelongToPath(files(), options.RootDir)
 		}
 	} else if options.ConfigFilePath != "" {
 		// If the rootDir is not specified, then the common source directory is the directory of the config file.
 		commonSourceDirectory = tspath.GetDirectoryPath(options.ConfigFilePath)
 		if checkSourceFilesBelongToPath != nil {
-			checkSourceFilesBelongToPath(files, commonSourceDirectory)
+			checkSourceFilesBelongToPath(files(), commonSourceDirectory)
 		}
 	} else {
-		commonSourceDirectory = computeCommonSourceDirectoryOfFilenames(files, currentDirectory, useCaseSensitiveFileNames)
+		commonSourceDirectory = computeCommonSourceDirectoryOfFilenames(files(), currentDirectory, useCaseSensitiveFileNames)
 	}
 
 	if len(commonSourceDirectory) > 0 {

--- a/internal/tsoptions/parsedcommandline.go
+++ b/internal/tsoptions/parsedcommandline.go
@@ -9,6 +9,7 @@ import (
 
 	"github.com/microsoft/typescript-go/internal/ast"
 	"github.com/microsoft/typescript-go/internal/core"
+	"github.com/microsoft/typescript-go/internal/diagnostics"
 	"github.com/microsoft/typescript-go/internal/glob"
 	"github.com/microsoft/typescript-go/internal/locale"
 	"github.com/microsoft/typescript-go/internal/module"
@@ -120,21 +121,33 @@ func (p *ParsedCommandLine) ParseInputOutputNames() {
 
 func (p *ParsedCommandLine) CommonSourceDirectory() string {
 	p.commonSourceDirectoryOnce.Do(func() {
+		files := core.Filter(p.ParsedConfig.FileNames, func(file string) bool {
+			return !(p.ParsedConfig.CompilerOptions.NoEmitForJsFiles.IsTrue() && tspath.HasJSFileExtension(file)) && !tspath.IsDeclarationFileName(file)
+		})
+
 		p.commonSourceDirectory = outputpaths.GetCommonSourceDirectory(
 			p.ParsedConfig.CompilerOptions,
-			func() []string {
-				return core.Filter(
-					p.ParsedConfig.FileNames,
-					func(file string) bool {
-						return !(p.ParsedConfig.CompilerOptions.NoEmitForJsFiles.IsTrue() && tspath.HasJSFileExtension(file)) &&
-							!tspath.IsDeclarationFileName(file)
-					})
-			},
+			files,
 			p.GetCurrentDirectory(),
 			p.UseCaseSensitiveFileNames(),
+			p.checkSourceFilesBelongToPath,
 		)
 	})
 	return p.commonSourceDirectory
+}
+
+func (p *ParsedCommandLine) checkSourceFilesBelongToPath(sourceFiles []string, rootDirectory string) bool {
+	allFilesBelongToPath := true
+	absoluteRootDirectoryPath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(rootDirectory, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
+	for _, file := range sourceFiles {
+		absoluteSourceFilePath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(file, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
+		if !strings.HasPrefix(absoluteSourceFilePath, absoluteRootDirectoryPath) {
+			p.Errors = append(p.Errors, ast.NewCompilerDiagnostic(diagnostics.File_0_is_not_under_rootDir_1_rootDir_is_expected_to_contain_all_source_files, file, rootDirectory))
+			allFilesBelongToPath = false
+		}
+	}
+
+	return allFilesBelongToPath
 }
 
 func (p *ParsedCommandLine) GetCurrentDirectory() string {

--- a/internal/tsoptions/parsedcommandline.go
+++ b/internal/tsoptions/parsedcommandline.go
@@ -121,9 +121,11 @@ func (p *ParsedCommandLine) ParseInputOutputNames() {
 
 func (p *ParsedCommandLine) CommonSourceDirectory() string {
 	p.commonSourceDirectoryOnce.Do(func() {
-		files := core.Filter(p.ParsedConfig.FileNames, func(file string) bool {
-			return !(p.ParsedConfig.CompilerOptions.NoEmitForJsFiles.IsTrue() && tspath.HasJSFileExtension(file)) && !tspath.IsDeclarationFileName(file)
-		})
+		files := func() []string {
+			return core.Filter(p.ParsedConfig.FileNames, func(file string) bool {
+				return !(p.ParsedConfig.CompilerOptions.NoEmitForJsFiles.IsTrue() && tspath.HasJSFileExtension(file)) && !tspath.IsDeclarationFileName(file)
+			})
+		}
 
 		p.commonSourceDirectory = outputpaths.GetCommonSourceDirectory(
 			p.ParsedConfig.CompilerOptions,

--- a/internal/tsoptions/parsedcommandline.go
+++ b/internal/tsoptions/parsedcommandline.go
@@ -138,11 +138,10 @@ func (p *ParsedCommandLine) CommonSourceDirectory() string {
 
 func (p *ParsedCommandLine) checkSourceFilesBelongToPath(sourceFiles []string, rootDirectory string) bool {
 	allFilesBelongToPath := true
-	absoluteRootDirectoryPath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(rootDirectory, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
 	for _, file := range sourceFiles {
 		absoluteSourceFilePath := tspath.GetCanonicalFileName(tspath.GetNormalizedAbsolutePath(file, p.GetCurrentDirectory()), p.UseCaseSensitiveFileNames())
-		if !strings.HasPrefix(absoluteSourceFilePath, absoluteRootDirectoryPath) {
-			p.Errors = append(p.Errors, ast.NewCompilerDiagnostic(diagnostics.File_0_is_not_under_rootDir_1_rootDir_is_expected_to_contain_all_source_files, file, rootDirectory))
+		if !tspath.ContainsPath(rootDirectory, file, p.comparePathsOptions) {
+			p.Errors = append(p.Errors, ast.NewCompilerDiagnostic(diagnostics.File_0_is_not_under_rootDir_1_rootDir_is_expected_to_contain_all_source_files, absoluteSourceFilePath, rootDirectory))
 			allFilesBelongToPath = false
 		}
 	}

--- a/testdata/baselines/reference/compiler/tsconfigRootdirInclude.errors.txt
+++ b/testdata/baselines/reference/compiler/tsconfigRootdirInclude.errors.txt
@@ -1,11 +1,11 @@
 error TS6059: File 'src1/main.ts' is not under 'rootDir' 'src2'. 'rootDir' is expected to contain all source files.
   The file is in the program because:
-    Matched by default include pattern '**/*'
+    Root file specified for compilation
 
 
 !!! error TS6059: File 'src1/main.ts' is not under 'rootDir' 'src2'. 'rootDir' is expected to contain all source files.
 !!! error TS6059:   The file is in the program because:
-!!! error TS6059:     Matched by default include pattern '**/*'
+!!! error TS6059:     Root file specified for compilation
 ==== tsconfig.json (0 errors) ====
     {
         "compilerOptions": {
@@ -13,8 +13,8 @@ error TS6059: File 'src1/main.ts' is not under 'rootDir' 'src2'. 'rootDir' is ex
             "strictNullChecks": true,
             "stableTypeOrdering": true,
             "rootDir": "./src2",
-            "include": ["src1/*"]
         }
+        "include": ["src1/*"]
     }
 ==== src1/main.ts (0 errors) ====
     

--- a/testdata/baselines/reference/compiler/tsconfigRootdirInclude.errors.txt
+++ b/testdata/baselines/reference/compiler/tsconfigRootdirInclude.errors.txt
@@ -1,0 +1,20 @@
+error TS6059: File 'src1/main.ts' is not under 'rootDir' 'src2'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+
+!!! error TS6059: File 'src1/main.ts' is not under 'rootDir' 'src2'. 'rootDir' is expected to contain all source files.
+!!! error TS6059:   The file is in the program because:
+!!! error TS6059:     Matched by default include pattern '**/*'
+==== tsconfig.json (0 errors) ====
+    {
+        "compilerOptions": {
+            "target": "es2020",
+            "strictNullChecks": true,
+            "stableTypeOrdering": true,
+            "rootDir": "./src2",
+            "include": ["src1/*"]
+        }
+    }
+==== src1/main.ts (0 errors) ====
+    

--- a/testdata/baselines/reference/compiler/tsconfigRootdirInclude.js
+++ b/testdata/baselines/reference/compiler/tsconfigRootdirInclude.js
@@ -1,0 +1,7 @@
+//// [tests/cases/compiler/tsconfigRootdirInclude.ts] ////
+
+//// [main.ts]
+
+
+//// [main.js]
+"use strict";

--- a/testdata/baselines/reference/compiler/tsconfigRootdirInclude.symbols
+++ b/testdata/baselines/reference/compiler/tsconfigRootdirInclude.symbols
@@ -1,0 +1,4 @@
+//// [tests/cases/compiler/tsconfigRootdirInclude.ts] ////
+
+=== src1/main.ts ===
+

--- a/testdata/baselines/reference/compiler/tsconfigRootdirInclude.types
+++ b/testdata/baselines/reference/compiler/tsconfigRootdirInclude.types
@@ -1,0 +1,4 @@
+//// [tests/cases/compiler/tsconfigRootdirInclude.ts] ////
+
+=== src1/main.ts ===
+

--- a/testdata/baselines/reference/tsbuild/outputPaths/when-rootDir-is-specified-but-not-all-files-belong-to-rootDir-and-is-composite.js
+++ b/testdata/baselines/reference/tsbuild/outputPaths/when-rootDir-is-specified-but-not-all-files-belong-to-rootDir-and-is-composite.js
@@ -15,7 +15,7 @@ export const x = 10;
 export type t = string;
 
 tsgo -b -v
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
 [[90mHH:MM:SS AM[0m] Projects in this build: 
     * tsconfig.json
@@ -23,6 +23,12 @@ Output::
 [[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is out of date because output file 'tsconfig.tsbuildinfo' does not exist
 
 [[90mHH:MM:SS AM[0m] Building project 'tsconfig.json'...
+
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
 
 //// [/home/src/tslibs/TS/Lib/lib.es2025.full.d.ts] *Lib*
 /// <reference no-default-lib="true"/>
@@ -54,10 +60,11 @@ export declare const x = 10;
 export const x = 10;
 
 //// [/home/src/workspaces/project/tsconfig.tsbuildinfo] *new* 
-{"version":"FakeTSVersion","root":[[2,3]],"fileNames":["lib.es2025.full.d.ts","./src/index.ts","./types/type.ts"],"fileInfos":[{"version":"8859c12c614ce56ba9a18e58384a198f-/// <reference no-default-lib=\"true\"/>\ninterface Boolean {}\ninterface Function {}\ninterface CallableFunction {}\ninterface NewableFunction {}\ninterface IArguments {}\ninterface Number { toExponential: any; }\ninterface Object {}\ninterface RegExp {}\ninterface String { charAt: any; }\ninterface Array<T> { length: number; [n: number]: T; }\ninterface ReadonlyArray<T> {}\ninterface SymbolConstructor {\n    (desc?: string | number): symbol;\n    for(name: string): symbol;\n    readonly toStringTag: symbol;\n}\ndeclare var Symbol: SymbolConstructor;\ninterface Symbol {\n    readonly [Symbol.toStringTag]: string;\n}\ndeclare const console: { log(msg: any): void; };","affectsGlobalScope":true,"impliedNodeFormat":1},{"version":"28e8748a7acd58f4f59388926e914f86-export const x = 10;","signature":"f9b4154a9a5944099ecf197d4519d083-export declare const x = 10;\n","impliedNodeFormat":1},{"version":"37c1beee3ff062c5ae875b0377f00093-export type t = string;","signature":"09b8756df8837e9d9dd668b82d3aad6c-export type t = string;\n","impliedNodeFormat":1}],"options":{"composite":true,"outDir":"./dist","rootDir":"./src"},"latestChangedDtsFile":"./types/type.d.ts"}
+{"version":"FakeTSVersion","errors":true,"root":[[2,3]],"fileNames":["lib.es2025.full.d.ts","./src/index.ts","./types/type.ts"],"fileInfos":[{"version":"8859c12c614ce56ba9a18e58384a198f-/// <reference no-default-lib=\"true\"/>\ninterface Boolean {}\ninterface Function {}\ninterface CallableFunction {}\ninterface NewableFunction {}\ninterface IArguments {}\ninterface Number { toExponential: any; }\ninterface Object {}\ninterface RegExp {}\ninterface String { charAt: any; }\ninterface Array<T> { length: number; [n: number]: T; }\ninterface ReadonlyArray<T> {}\ninterface SymbolConstructor {\n    (desc?: string | number): symbol;\n    for(name: string): symbol;\n    readonly toStringTag: symbol;\n}\ndeclare var Symbol: SymbolConstructor;\ninterface Symbol {\n    readonly [Symbol.toStringTag]: string;\n}\ndeclare const console: { log(msg: any): void; };","affectsGlobalScope":true,"impliedNodeFormat":1},{"version":"28e8748a7acd58f4f59388926e914f86-export const x = 10;","signature":"f9b4154a9a5944099ecf197d4519d083-export declare const x = 10;\n","impliedNodeFormat":1},{"version":"37c1beee3ff062c5ae875b0377f00093-export type t = string;","signature":"09b8756df8837e9d9dd668b82d3aad6c-export type t = string;\n","impliedNodeFormat":1}],"options":{"composite":true,"outDir":"./dist","rootDir":"./src"},"semanticDiagnosticsPerFile":[1,2,3],"latestChangedDtsFile":"./types/type.d.ts"}
 //// [/home/src/workspaces/project/tsconfig.tsbuildinfo.readable.baseline.txt] *new* 
 {
   "version": "FakeTSVersion",
+  "errors": true,
   "root": [
     {
       "files": [
@@ -116,8 +123,13 @@ export const x = 10;
     "outDir": "./dist",
     "rootDir": "./src"
   },
+  "semanticDiagnosticsPerFile": [
+    "lib.es2025.full.d.ts",
+    "./src/index.ts",
+    "./types/type.ts"
+  ],
   "latestChangedDtsFile": "./types/type.d.ts",
-  "size": 1340
+  "size": 1391
 }
 //// [/home/src/workspaces/project/types/type.d.ts] *new* 
 export type t = string;
@@ -128,9 +140,9 @@ export {};
 
 tsconfig.json::
 SemanticDiagnostics::
-*refresh*    /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
-*refresh*    /home/src/workspaces/project/src/index.ts
-*refresh*    /home/src/workspaces/project/types/type.ts
+*not cached* /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
+*not cached* /home/src/workspaces/project/src/index.ts
+*not cached* /home/src/workspaces/project/types/type.ts
 Signatures::
 (stored at emit) /home/src/workspaces/project/src/index.ts
 (stored at emit) /home/src/workspaces/project/types/type.ts
@@ -139,22 +151,45 @@ Signatures::
 Edit [0]:: no change
 
 tsgo -b -v
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
 [[90mHH:MM:SS AM[0m] Projects in this build: 
     * tsconfig.json
 
-[[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is up to date because newest input 'types/type.ts' is older than output 'tsconfig.tsbuildinfo'
+[[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is out of date because buildinfo file 'tsconfig.tsbuildinfo' indicates that program needs to report errors.
+
+[[90mHH:MM:SS AM[0m] Building project 'tsconfig.json'...
+
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
 
 
+tsconfig.json::
+SemanticDiagnostics::
+*not cached* /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
+*not cached* /home/src/workspaces/project/src/index.ts
+*not cached* /home/src/workspaces/project/types/type.ts
+Signatures::
 
 
 Edit [1]:: Normal build without change, that does not block emit on error to show files that get emitted
 
 tsgo -p /home/src/workspaces/project/tsconfig.json
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
+
 
 tsconfig.json::
 SemanticDiagnostics::
+*not cached* /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
+*not cached* /home/src/workspaces/project/src/index.ts
+*not cached* /home/src/workspaces/project/types/type.ts
 Signatures::

--- a/testdata/baselines/reference/tsbuild/outputPaths/when-rootDir-is-specified-but-not-all-files-belong-to-rootDir.js
+++ b/testdata/baselines/reference/tsbuild/outputPaths/when-rootDir-is-specified-but-not-all-files-belong-to-rootDir.js
@@ -14,7 +14,7 @@ export const x = 10;
 export type t = string;
 
 tsgo -b -v
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
 [[90mHH:MM:SS AM[0m] Projects in this build: 
     * tsconfig.json
@@ -22,6 +22,12 @@ Output::
 [[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is out of date because output file 'tsconfig.tsbuildinfo' does not exist
 
 [[90mHH:MM:SS AM[0m] Building project 'tsconfig.json'...
+
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
 
 //// [/home/src/tslibs/TS/Lib/lib.es2025.full.d.ts] *Lib*
 /// <reference no-default-lib="true"/>
@@ -50,10 +56,11 @@ declare const console: { log(msg: any): void; };
 export const x = 10;
 
 //// [/home/src/workspaces/project/tsconfig.tsbuildinfo] *new* 
-{"version":"FakeTSVersion","root":["./src/index.ts","./types/type.ts"]}
+{"version":"FakeTSVersion","errors":true,"root":["./src/index.ts","./types/type.ts"]}
 //// [/home/src/workspaces/project/tsconfig.tsbuildinfo.readable.baseline.txt] *new* 
 {
   "version": "FakeTSVersion",
+  "errors": true,
   "root": [
     {
       "files": [
@@ -68,7 +75,7 @@ export const x = 10;
       "original": "./types/type.ts"
     }
   ],
-  "size": 71
+  "size": 85
 }
 //// [/home/src/workspaces/project/types/type.js] *new* 
 export {};
@@ -76,30 +83,54 @@ export {};
 
 tsconfig.json::
 SemanticDiagnostics::
-*refresh*    /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
-*refresh*    /home/src/workspaces/project/src/index.ts
-*refresh*    /home/src/workspaces/project/types/type.ts
+*not cached* /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
+*not cached* /home/src/workspaces/project/src/index.ts
+*not cached* /home/src/workspaces/project/types/type.ts
 Signatures::
 
 
 Edit [0]:: no change
 
 tsgo -b -v
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
 [[90mHH:MM:SS AM[0m] Projects in this build: 
     * tsconfig.json
 
-[[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is up to date because newest input 'types/type.ts' is older than output 'dist/index.js'
+[[90mHH:MM:SS AM[0m] Project 'tsconfig.json' is out of date because buildinfo file 'tsconfig.tsbuildinfo' indicates that program needs to report errors.
 
+[[90mHH:MM:SS AM[0m] Building project 'tsconfig.json'...
 
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
+
+//// [/home/src/workspaces/project/dist/index.js] *rewrite with same content*
+//// [/home/src/workspaces/project/tsconfig.tsbuildinfo] *rewrite with same content*
+//// [/home/src/workspaces/project/tsconfig.tsbuildinfo.readable.baseline.txt] *rewrite with same content*
+//// [/home/src/workspaces/project/types/type.js] *rewrite with same content*
+
+tsconfig.json::
+SemanticDiagnostics::
+*not cached* /home/src/tslibs/TS/Lib/lib.es2025.full.d.ts
+*not cached* /home/src/workspaces/project/src/index.ts
+*not cached* /home/src/workspaces/project/types/type.ts
+Signatures::
 
 
 Edit [1]:: Normal build without change, that does not block emit on error to show files that get emitted
 
 tsgo -p /home/src/workspaces/project/tsconfig.json
-ExitStatus:: Success
+ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
+[91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/types/type.ts' is not under 'rootDir' '/home/src/workspaces/project/src'. 'rootDir' is expected to contain all source files.
+  The file is in the program because:
+    Matched by default include pattern '**/*'
+
+Found 1 error.
+
 //// [/home/src/workspaces/project/dist/index.js] *rewrite with same content*
 //// [/home/src/workspaces/project/types/type.js] *rewrite with same content*
 

--- a/testdata/baselines/reference/tsc/projectReferences/errors-when-a-file-is-outside-the-rootdir.js
+++ b/testdata/baselines/reference/tsc/projectReferences/errors-when-a-file-is-outside-the-rootdir.js
@@ -17,13 +17,18 @@ export { }
 tsgo --p alpha/tsconfig.json
 ExitStatus:: DiagnosticsPresent_OutputsGenerated
 Output::
+[96malpha/src/a.ts[0m:[93m1[0m:[93m20[0m - [91merror[0m[90m TS6059: [0mFile '/home/src/workspaces/project/beta/b.ts' is not under 'rootDir' '/home/src/workspaces/project/alpha'. 'rootDir' is expected to contain all source files.
+
+[7m1[0m import * as b from '../../beta/b'
+[7m [0m [91m                   ~~~~~~~~~~~~~~[0m
+
 [96malpha/src/a.ts[0m:[93m1[0m:[93m20[0m - [91merror[0m[90m TS6307: [0mFile '/home/src/workspaces/project/beta/b.ts' is not listed within the file list of project '/home/src/workspaces/project/alpha/tsconfig.json'. Projects must list all files or use an 'include' pattern.
 
 [7m1[0m import * as b from '../../beta/b'
 [7m [0m [91m                   ~~~~~~~~~~~~~~[0m
 
 
-Found 1 error in alpha/src/a.ts[90m:1[0m
+Found 2 errors in the same file, starting at: alpha/src/a.ts[90m:1[0m
 
 //// [/home/src/tslibs/TS/Lib/lib.es2025.full.d.ts] *Lib*
 /// <reference no-default-lib="true"/>

--- a/testdata/tests/cases/compiler/tsconfigRootdirInclude.ts
+++ b/testdata/tests/cases/compiler/tsconfigRootdirInclude.ts
@@ -5,8 +5,8 @@
         "strictNullChecks": true,
         "stableTypeOrdering": true,
         "rootDir": "./src2",
-        "include": ["src1/*"]
     }
+    "include": ["src1/*"]
 }
 // @filename: src1/main.ts
 

--- a/testdata/tests/cases/compiler/tsconfigRootdirInclude.ts
+++ b/testdata/tests/cases/compiler/tsconfigRootdirInclude.ts
@@ -1,0 +1,13 @@
+// @filename: tsconfig.json
+{
+    "compilerOptions": {
+        "target": "es2020",
+        "strictNullChecks": true,
+        "stableTypeOrdering": true,
+        "rootDir": "./src2",
+        "include": ["src1/*"]
+    }
+}
+// @filename: src1/main.ts
+
+

